base/client.jl

```diff
@@ -84,35 +84,65 @@ answer_color() = text_colors[repl_color("JULIA_ANSWER_COLOR", default_color_answ
 stackframe_lineinfo_color() = repl_color("JULIA_STACKFRAME_LINEINFO_COLOR", :bold)
 stackframe_function_color() = repl_color("JULIA_STACKFRAME_FUNCTION_COLOR", :bold)
 
-function repl_cmd(cmd, out)
-    shell = shell_split(get(ENV,"JULIA_SHELL",get(ENV,"SHELL","/bin/sh")))
-    shell_name = Base.basename(shell[1])
+repl_cmd(line::AbstractString, out) = repl_cmd(Val(Sys.iswindows()), line, out)
 
-    if isempty(cmd.exec)
+function repl_cmd(iswindows::Val{false}, line::AbstractString, out)
+    shell = shell_split(get(ENV, "JULIA_SHELL", get(ENV,"SHELL","/bin/sh")))
+    shell_name = basename(shell[1])
+    cmd = cmd_gen(eval(Main, shell_parse(line)[1]))
+    if isempty(cmd)
         throw(ArgumentError("no cmd to execute"))
-    elseif cmd.exec[1] == "cd"
-        new_oldpwd = pwd()
-        if length(cmd.exec) > 2
-            throw(ArgumentError("cd method only takes one argument"))
-        elseif length(cmd.exec) == 2
-            dir = cmd.exec[2]
-            if dir == "-"
-                if !haskey(ENV, "OLDPWD")
-                    error("cd: OLDPWD not set")
-                end
-                cd(ENV["OLDPWD"])
-            else
-                cd(@static Sys.iswindows() ? dir : readchomp(`$shell -c "echo $(shell_escape(dir))"`))
+    elseif cmd[1] == "cd" && length(cmd) <= 2
+        repl_cd(cmd, shell, out)
+    else
+        run(ignorestatus(isa(STDIN, TTY) ? `$shell -i -c "$(shell_wrap_true(shell_name, cmd))"` : `$shell -c "$(shell_wrap_true(shell_name, cmd))"`))
```

Comment:
This could be written a little more cleanly as
```julia
ttyopt = STDIN isa TTY ? "-i" : ""
run(ignorestatus(`$shell $ttyopt -c "$(shell_wrap_true(shell_name, cmd))"`))
```
since `-i` is the only thing that differs between the two branches.

Reply:
Yes of course, will update.

Although probably better to use command syntax rather strings:
```
ttyopt = STDIN isa TTY ? `-i` : ``
```
https://discourse.julialang.org/t/interpolating-an-empty-string-inside-backticks/2428/3

Reply:
Ah yes, good call. I had forgotten about that.